File: src/main/java/com/libraryjava/service/UserService.java
package com.libraryjava.service;

import com.libraryjava.domain.user.User;
import com.libraryjava.domain.user.loanhistory.UserLoanStatus;
import com.libraryjava.dto.user.UserMakeDto;
import com.libraryjava.dto.user.UserResponseDto;
import com.libraryjava.dto.user.UserUpdateDto;
import com.libraryjava.dto.user.response.BookHistoryResponse;
import com.libraryjava.dto.user.response.UserLoanHistoryResponse;
import com.libraryjava.repository.BookRepository;
import com.libraryjava.repository.UserLoanHistoryRepository;
import com.libraryjava.repository.UserRepository;
import jakarta.persistence.EntityManager;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
@Slf4j
public class UserService {
    private final UserLoanHistoryRepository userLoanHistoryRepository;
    private final BookRepository bookRepository;
    private final UserRepository userRepository;

    public List<UserResponseDto> getUsers() {
        return userRepository.findAll().stream()
                .map(UserResponseDto::of)
                .toList();
    }

    @Transactional
    public void makeUser(UserMakeDto userMakeDto) {
        userRepository.save(new User(userMakeDto.name(), userMakeDto.age()));
    }

    @Transactional
    public void updateUser(UserUpdateDto userUpdateDto) {
        User findUser = userRepository.findById(userUpdateDto.id())
                .orElseThrow(IllegalArgumentException::new);

        findUser.updateName(userUpdateDto.name());
    }

    @Transactional
    public void deleteUser(String name) {
        User findUser = userRepository.findByName(name)
                .orElseThrow(IllegalArgumentException::new);

        userRepository.delete(findUser);
    }

    public List<UserLoanHistoryResponse> getUserLoanHistories() {
        return userRepository.findAllWithHistories().stream()
                .map(user -> UserLoanHistoryResponse.of(user.getName(),
                        user.getLoanHistories().stream()
                                .map(history -> BookHistoryResponse.of(history.getBookName(), history.getUserLoanStatus() == UserLoanStatus.RETURNED))
                                .toList()))
                .toList();
    }
}

File: src/main/kotlin/com/library/service/user/UserService.kt
package com.library.service.user

import com.library.domain.user.User
import com.library.domain.user.UserRepository
import com.library.domain.user.UserStatus
import com.library.domain.user.loanhistory.UserLoanStatus
import com.library.dto.user.*
import com.library.util.fail
import com.library.util.findByIdOrThrow
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class UserService(
    private val userRepository: UserRepository
) {

    @Transactional
    fun saveUser(request: UserCreateRequest) {
        val newUser = User(request.name, request.age, UserStatus.ACTIVE)
        userRepository.save(newUser)
    }

    fun getUsers(): List<UserResponse> {
        return userRepository.findAll()
            .map { user -> UserResponse.of(user) }
    }

    @Transactional
    fun updateUserName(request: UserUpdateRequest) {
        val user = userRepository.findByIdOrThrow(request.id)
        user.updateName(request.name)
    }

    @Transactional
    fun deleteUser(name: String) {
        val user = userRepository.findByName(name) ?: fail()
        userRepository.delete(user)
    }

    fun getUserLoanHistories(): List<UserLoanHistoryResponse> {
        return userRepository.findAllWithHistories().map {
            user -> UserLoanHistoryResponse(
                name = user.name,
                books = user.userLoanHistories.map {
                    history -> BookHistoryResponse(
                        name = history.bookName,
                        isReturn = history.status == UserLoanStatus.RETURNED,
                    )
                }
            )
        }
    }
}